CRM/Core/BAO/ActionSchedule.php

```diff
@@ -267,44 +267,51 @@ public static function sendMailings($mappingID, $now) {
       );
 
       $multilingual = CRM_Core_I18n::isMultilingual();
+      $tokenProcessor = self::createTokenProcessor($actionSchedule, $mapping);
       while ($dao->fetch()) {
-        $errors = [];
-        try {
-          $tokenProcessor = self::createTokenProcessor($actionSchedule, $mapping);
-          $row = $tokenProcessor->addRow()
-            ->context('contactId', $dao->contactID)
-            ->context('actionSearchResult', (object) $dao->toArray());
-
-          // switch language if necessary
-          if ($multilingual) {
-            $preferred_language = CRM_Core_DAO::getFieldValue('CRM_Contact_DAO_Contact', $dao->contactID, 'preferred_language');
-            $row->context('locale', CRM_Core_BAO_ActionSchedule::pickLocale($actionSchedule->communication_language, $preferred_language));
+        $row = $tokenProcessor->addRow()
+          ->context('contactId', $dao->contactID)
+          ->context('actionSearchResult', (object) $dao->toArray());
+
+        // switch language if necessary
+        if ($multilingual) {
+          $preferred_language = CRM_Core_DAO::getFieldValue('CRM_Contact_DAO_Contact', $dao->contactID, 'preferred_language');
+          $row->context('locale', CRM_Core_BAO_ActionSchedule::pickLocale($actionSchedule->communication_language, $preferred_language));
+        }
+
+        foreach ($dao->toArray() as $key => $value) {
+          if (preg_match('/^tokenContext_(.*)/', $key, $m)) {
```

Comment:
So really this is the meaningful change in here - I think some comments around it would help. The fact this still passes tests means it must still work for contribution tokens - but I'm quite sure how

Reply:
@eileenmcnaughton Agree about comments. Added PR for that: #21098

For contribution tokens, this PR doesn't actually change anything. `CRM_Contribution_Tokens::alterActionScheduleQuery()` still does the same prefetch that it did before. That'll be the subsequent PR.

For this PR, I only touched `CRM_Activity_Tokens::alterActionScheduleQuery()` because that was the simplest one - a way to demonstrate the mechanism.